# Project 2: Data Profiling and Preparation

### Project Objectives
> Analyze business data using methods to determine the nature of the data for cleansing purposes <br>
> Practice how to clean data to address missing, bad, or inaccurate data <br>
> Practice data accessing to implement dataset importing and exporting

##### Project 2: Data Profiling and Preparation
MISM 6205: Data Wrangling for Business <br>
Prof. Tareq Nasralah <br>
Austin Roth-Eagle, David Martin, Zachary Errichetti

## 1. Conduct Data Access and Profiling
1. To begin, read the survey dataset into a pandas DataFrame.
2. Then, create a subset of the DataFrame to include only passenger satisfaction data about the Business Class. Name the created DataFrame subset “df_Business”.
3. Collect information about the “df_Business” dataset and its validity. Report the following: 

> 1. The number of observations (or rows) and the number of variables (or columns) in the dataset.
2. The name and data type of each data column.
3. The unique values of each column.
4. The number of missing values of each column.
5. Summary statistics for each data column.
6. Frequency distribution (list the number of times each unique value appears) for each data column.
7. The number of fully duplicated data rows.
            
Based on the information that you collected, provide a list of the issues that the data contains as an annotation

### Read the survey dataset into a pandas DataFrame.

In [1]:
import pandas as pd
import numpy as np

In [2]:
pd.set_option('mode.chained_assignment',None)

In [3]:
df = pd.read_csv('Passengers_Satisfaction_Survey (1).csv')
df

Gender      Customer Type   Age   Type of Travel     Class  \
0         Male     Loyal Customer  69.0  Personal Travel  Business   
1         Male  Disloyal Customer  20.0  Business travel       Eco   
2         Male  Disloyal Customer  41.0  Business travel  Business   
3         Male     Loyal Customer  52.0  Business travel  Business   
4       Female  Disloyal Customer  21.0  Business travel  Business   
...        ...                ...   ...              ...       ...   
104905    Male     Loyal Customer  45.0  Business travel  Business   
104906  Female     Loyal Customer  35.0  Business travel  Business   
104907    Male     Loyal Customer  47.0  Business travel  Business   
104908    Male     Loyal Customer  42.0  Business travel  Business   
104909  Female           disloyal   0.0  Business travel  Business   

        Flight Distance  Inflight wifi service  \
0               -5421.0                    3.0   
1                   0.0                    3.0   
2                   0.0                    1.0   
3                   0.0                    2.0   
4                   0.0                    4.0   
...                 ...                    ...   
104905           3305.0                    2.0   
104906            337.0                    5.0   
104907           3496.0                    2.0   
104908           3605.0                    4.0   
104909            862.0                    5.0   

        Departure/Arrival time convenient Ease of Online booking  \
0                                     2.0                      3   
1                                     3.0                      3   
2                                     1.0                    NaN   
3                                     4.0                      4   
4                                     0.0                    NaN   
...                                   ...                    ...   
104905                                2.0                      2   
104906                                5.0                      5   
104907                                2.0                      2   
104908                                4.0                      4   
104909                                5.0                      5   

       Food and drink  Inflight entertainment Baggage handling  Cleanliness  \
0                   5                     1.0                1          3.0   
1                   2                    10.0                4          2.0   
2                 NaN                     NaN              NaN          NaN   
3                   1                     2.0                2          4.0   
4                 NaN                     NaN              NaN          NaN   
...               ...                     ...              ...          ...   
104905              3                     4.0                4          3.0   
104906              2                     4.0                4          5.0   
104907              2                     5.0                5          4.0   
104908              3                     4.0                4          3.0   
104909              5                     5.0                1          5.0   

        Departure Delay in Minutes  Arrival Delay in Minutes  Satisfaction  
0                             14.0                       0.0  Dissatisfied  
1                              0.0                       0.0  dissatisfied  
2                              NaN                       NaN       neutral  
3                              0.0                       6.0  Dissatisfied  
4                              NaN                       NaN       neutral  
...                            ...                       ...           ...  
104905                         8.0                       0.0     satisfied  
104906                        54.0                      40.0     satisfied  
104907                         0.0                       2.0     satisfied  
104908                        

### Create a subset of the DataFrame to include only passenger satisfaction data about the Business Class. Name the created DataFrame subset “df_Business”.

In [4]:
df_Business = df[(df["Class"] == "Business")]
df_Business

Gender      Customer Type   Age   Type of Travel     Class  \
0         Male     Loyal Customer  69.0  Personal Travel  Business   
2         Male  Disloyal Customer  41.0  Business travel  Business   
3         Male     Loyal Customer  52.0  Business travel  Business   
4       Female  Disloyal Customer  21.0  Business travel  Business   
5       Female     Loyal Customer  57.0  Business travel  Business   
...        ...                ...   ...              ...       ...   
104905    Male     Loyal Customer  45.0  Business travel  Business   
104906  Female     Loyal Customer  35.0  Business travel  Business   
104907    Male     Loyal Customer  47.0  Business travel  Business   
104908    Male     Loyal Customer  42.0  Business travel  Business   
104909  Female           disloyal   0.0  Business travel  Business   

        Flight Distance  Inflight wifi service  \
0               -5421.0                    3.0   
2                   0.0                    1.0   
3                   0.0                    2.0   
4                   0.0                    4.0   
5                   0.0                    3.0   
...                 ...                    ...   
104905           3305.0                    2.0   
104906            337.0                    5.0   
104907           3496.0                    2.0   
104908           3605.0                    4.0   
104909            862.0                    5.0   

        Departure/Arrival time convenient Ease of Online booking  \
0                                     2.0                      3   
2                                     1.0                    NaN   
3                                     4.0                      4   
4                                     0.0                    NaN   
5                                     4.0                      4   
...                                   ...                    ...   
104905                                2.0                      2   
104906                                5.0                      5   
104907                                2.0                      2   
104908                                4.0                      4   
104909                                5.0                      5   

       Food and drink  Inflight entertainment Baggage handling  Cleanliness  \
0                   5                     1.0                1          3.0   
2                 NaN                     NaN              NaN          NaN   
3                   1                     2.0                2          4.0   
4                 NaN                     NaN              NaN          NaN   
5                   5                     2.0                3          2.0   
...               ...                     ...              ...          ...   
104905              3                     4.0                4          3.0   
104906              2                     4.0                4          5.0   
104907              2                     5.0                5          4.0   
104908              3                     4.0                4          3.0   
104909              5                     5.0                1          5.0   

        Departure Delay in Minutes  Arrival Delay in Minutes  Satisfaction  
0                             14.0                       0.0  Dissatisfied  
2                              NaN                       NaN       neutral  
3                              0.0                       6.0  Dissatisfied  
4                              NaN                       NaN       neutral  
5                              0.0                       0.0  Dissatisfied  
...                            ...                       ...           ...  
104905                         8.0                       0.0     satisfied  
104906                        54.0                      40.0     satisfied  
104907                         0.0                       2.0     satisfied  
104908                        

### 1. The number of observations (or rows) and the number of variables (or columns) in the dataset.

In [5]:
df_Business.shape

(50000, 16)

### 2. The name and data type of each data column.

In [6]:
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 104909
Data columns (total 16 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Gender                             48239 non-null  object 
 1   Customer Type                      48239 non-null  object 
 2   Age                                47373 non-null  float64
 3   Type of Travel                     48239 non-null  object 
 4   Class                              50000 non-null  object 
 5   Flight Distance                    46604 non-null  float64
 6   Inflight wifi service              48239 non-null  float64
 7   Departure/Arrival time convenient  48239 non-null  float64
 8   Ease of Online booking             42705 non-null  object 
 9   Food and drink                     42705 non-null  object 
 10  Inflight entertainment             43649 non-null  float64
 11  Baggage handling                   43649 non-null  ob

#### Each Column name and Dataype:

|  Column                            | Dtype  
|  :------                           |   -----:
|  Gender                            |   object  
|  Customer Type                     |   object 
|  Age                               |   float64 
|  Type of Travel                    |   object 
|  Class                             |   object 
|  Flight Distance                   |   float64
|  Inflight wifi service             |   float64 
|  Departure/Arrival time convenient |   float64  
|  Ease of Online booking            |   object 
|  Food and drink                    |   object  
|  Inflight entertainment            |   float64 
|  Baggage handling                  |   object  
|  Cleanliness                       |   float64 
|  Departure Delay in Minutes        |   float64 
|  Arrival Delay in Minutes          |   float64  
|  Satisfaction                      |   object  

### 3. The unique values of each column.

In [7]:
df_Business.columns

Index(['Gender', 'Customer Type', 'Age', 'Type of Travel', 'Class',
       'Flight Distance', 'Inflight wifi service',
       'Departure/Arrival time convenient', 'Ease of Online booking',
       'Food and drink', 'Inflight entertainment', 'Baggage handling',
       'Cleanliness', 'Departure Delay in Minutes', 'Arrival Delay in Minutes',
       'Satisfaction'],
      dtype='object')

In [8]:
cols = df_Business.columns

# for each column
for col in cols:
    print(col)

    # get a list of unique values
    unique = df_Business[col].unique()
    print(unique, '\n---------------------------------------\n\n')

Gender
['Male' 'Female' 'F' 'M' nan] 
---------------------------------------


Customer Type
['Loyal Customer' 'Disloyal Customer' nan 'disloyal' 'Loyal'] 
---------------------------------------


Age
[ 69.  41.  52.  21.  57.  45.  39.  33.  10.  55.  59.  53.  56.  47.
  54.  25.  51.  60.  27.  43.  20.  19.  63.  50.  nan  38.  46.  85.
  26.  42.  30.  24.  22.  17.  48.  40.  16.  32.  35.  44.  49.  65.
  67.  66.  64.   7.  37.  36.  15.  62.  34.   0.  68.  70.  58.  11.
  74.  13.  80.  75.   9.  61.  71.  12.  73.  72.  18.  79.   8.  76.
  14.  77.  78. 180. 185.] 
---------------------------------------


Type of Travel
['Personal Travel' 'Business travel' 'Business' nan] 
---------------------------------------


Class
['Business'] 
---------------------------------------


Flight Distance
[-5.421e+03  0.000e+00  5.600e+01 ...  4.983e+03  2.820e+06  9.000e+06] 
---------------------------------------


Inflight wifi service
[ 3.  1.  2.  4.  0.  5. nan] 
---------------

### 4. The number of missing values of each column.

In [9]:
df_Business.isnull().sum()

Gender                               1761
Customer Type                        1761
Age                                  2627
Type of Travel                       1761
Class                                   0
Flight Distance                      3396
Inflight wifi service                1761
Departure/Arrival time convenient    1761
Ease of Online booking               7295
Food and drink                       7295
Inflight entertainment               6351
Baggage handling                     6351
Cleanliness                          6351
Departure Delay in Minutes           6351
Arrival Delay in Minutes             7902
Satisfaction                         2160
dtype: int64

### 5. Summary statistics for each data column.

In [10]:
df_Business.describe(include='all')

Gender   Customer Type           Age   Type of Travel     Class  \
count    48239           48239  47373.000000            48239     50000   
unique       4               4           NaN                3         1   
top     Female  Loyal Customer           NaN  Business travel  Business   
freq     23001           40770           NaN            44063     50000   
mean       NaN             NaN     41.853186              NaN       NaN   
std        NaN             NaN     13.730683              NaN       NaN   
min        NaN             NaN      0.000000              NaN       NaN   
25%        NaN             NaN     34.000000              NaN       NaN   
50%        NaN             NaN     43.000000              NaN       NaN   
75%        NaN             NaN     52.000000              NaN       NaN   
max        NaN             NaN    185.000000              NaN       NaN   

        Flight Distance  Inflight wifi service  \
count      4.660400e+04           48239.000000   
unique              NaN                    NaN   
top                 NaN                    NaN   
freq                NaN                    NaN   
mean       1.934737e+03               2.774519   
std        4.369344e+04               1.423917   
min       -5.421000e+03               0.000000   
25%        5.940000e+02               2.000000   
50%        1.597000e+03               3.000000   
75%        2.565000e+03               4.000000   
max        9.000000e+06               5.000000   

        Departure/Arrival time convenient Ease of Online booking  \
count                        48239.000000                  42705   
unique                                NaN                      7   
top                                   NaN                      3   
freq                                  NaN                   8635   
mean                             2.910695                    NaN   
std                              1.501711                    NaN   
min                              0.000000                    NaN   
25%                              2.000000                    NaN   
50%                              3.000000                    NaN   
75%                              4.000000                    NaN   
max                              5.000000                    NaN   

       Food and drink  Inflight entertainment Baggage handling   Cleanliness  \
count           42705            43649.000000            43649  43649.000000   
unique              7                     NaN                6           NaN   
top                 4                     NaN                4           NaN   
freq            10890                     NaN            17056           NaN   
mean              NaN                3.835391              NaN      3.737520   
std               NaN                1.233508              NaN      0.987664   
min               NaN                0.000000              NaN      0.000000   
25%               NaN                3.000000              NaN      3.000000   
50%               NaN                4.000000              NaN      4.000000   
75%               NaN                5.000000              NaN      5.000000   
max               NaN               10.000000              NaN      5.000000   

        Departure Delay in Minutes  Arrival Delay in Minutes Satisfaction  
count                 43649.000000              42098.000000        47840  
unique                         NaN                       NaN            4  
top                            NaN                       NaN    satisfied  
freq                           NaN                       NaN        30400  
mean                     13.595363                 13.059647          NaN  
std                      36.595048                 38.116530          NaN  
min                      -2.000000                -45.000000          NaN  
25%                       0.000000                  0.000000          NaN  
50%                       0.000000          

### 6. Frequency distribution (list the number of times each unique value appears) for each data column.    

In [11]:
cols = df.columns
for col in cols:
    print(col)
    value_counts = df[col].value_counts()
    print(value_counts, '\n\n-------------------------------------------------')

Gender
Female    49211
Male      47963
M          2055
F          1361
Name: Gender, dtype: int64 

-------------------------------------------------
Customer Type
Loyal Customer       81964
Disloyal Customer    16891
disloyal              1730
Loyal                    5
Name: Customer Type, dtype: int64 

-------------------------------------------------
Age
39.0     2995
25.0     2829
40.0     2596
44.0     2508
41.0     2482
         ... 
78.0       35
85.0       17
150.0       2
180.0       1
185.0       1
Name: Age, Length: 76, dtype: int64 

-------------------------------------------------
Type of Travel
Business travel    64845
Personal Travel    30616
Business            5129
Name: Type of Travel, dtype: int64 

-------------------------------------------------
Class
Business    50000
Eco         47313
Eco Plus     7597
Name: Class, dtype: int64 

-------------------------------------------------
Flight Distance
337.0        603
404.0        372
594.0        352
447.0        3

### 7. The number of fully duplicated data rows.

In [12]:
df_Business.duplicated()

0         False
2         False
3         False
4         False
5         False
          ...  
104905    False
104906    False
104907    False
104908    False
104909    False
Length: 50000, dtype: bool

In [13]:
# Selecting duplicate based on all columns
duplicate = df_Business[df_Business.duplicated()]
  
print("Duplicate Rows :")
duplicate

Duplicate Rows :


Gender      Customer Type   Age   Type of Travel     Class  \
153          F  Disloyal Customer  25.0         Business  Business   
349        NaN                NaN   NaN              NaN  Business   
369        NaN                NaN   NaN              NaN  Business   
705        NaN                NaN   NaN              NaN  Business   
1159    Female     Loyal Customer  49.0  Business travel  Business   
...        ...                ...   ...              ...       ...   
104737  Female     Loyal Customer  37.0  Business travel  Business   
104768     NaN                NaN   NaN              NaN  Business   
104789    Male     Loyal Customer  24.0         Business  Business   
104825       M     Loyal Customer  30.0  Business travel  Business   
104840       M     Loyal Customer  30.0  Business travel  Business   

        Flight Distance  Inflight wifi service  \
153                67.0                    1.0   
349                 NaN                    NaN   
369                 NaN                    NaN   
705                 NaN                    NaN   
1159              101.0                    1.0   
...                 ...                    ...   
104737           3987.0                    3.0   
104768              NaN                    NaN   
104789           3993.0                    2.0   
104825              NaN                    3.0   
104840              NaN                    4.0   

        Departure/Arrival time convenient Ease of Online booking  \
153                                   0.0                      1   
349                                   NaN                    NaN   
369                                   NaN                    NaN   
705                                   NaN                    NaN   
1159                                  4.0                      4   
...                                   ...                    ...   
104737                                1.0                      3   
104768                                NaN                    NaN   
104789                                5.0                      5   
104825                                3.0                    NaN   
104840                                4.0                    NaN   

       Food and drink  Inflight entertainment Baggage handling  Cleanliness  \
153                 3                     3.0                4          3.0   
349               NaN                     NaN              NaN          NaN   
369               NaN                     NaN              NaN          NaN   
705               NaN                     NaN              NaN          NaN   
1159                3                     2.0                1          3.0   
...               ...                     ...              ...          ...   
104737              5                     5.0                5          5.0   
104768            NaN                     NaN              NaN          NaN   
104789              2                     2.0                A          2.0   
104825            NaN                     4.0                1          4.0   
104840            NaN                     4.0                4          4.0   

        Departure Delay in Minutes  Arrival Delay in Minutes  Satisfaction  
153                            0.0                       0.0           NaN  
349                            NaN                       NaN  dissatisfied  
369                            NaN                       NaN  dissatisfied  
705                            NaN                       NaN  dissatisfied  
1159                          11.0                      14.0  Dissatisfied  
...                            ...                       ...           ...  
104737                        10.0                      23.0     satisfied  
104768                         NaN                       NaN  dissatisfied  
104789                        23.0                     -45.0  dissatisfied  
104825                        

# Summary

### Based on the information that you collected, provide a list of the issues that the data contains as an annotation.

There are inconsistencies in how data is entered in each column, for example:
> The "Gender" column includes "Female", "Male", "F", and "M". Remapping F to Female and M to Male (or vise-versa) would fix this issue. <br>

> The "Customer Type" column includes "Loyal Customer", "Disloyal Customer", "Disloyal", and "Loyal". Remapping values to just Loyal/Disloyal is needed (as well as correct inconsistent capitalization). <br>

> The "Type of Travel" column has both  "Business travel", and "Business". These need to be consolidated and have consistent capitalization. <br>

> There are many flights where the "age" is 0. It is unlikely toddlers are flying business class, and should either be confirmed as correct ages, or replaced with NaN (if the individual did not wish to disclose their age) <br>

> The "Flight Distance" column includes negative values, which is not possible and is likely an error to be removed (or confirm if the value should be positive). <br>

> "Baggage handling" has a number of rows with the value "A" instead of numeric ratings. These should be dropped, remapped, or made to be nulls.

> "Food and drink" have the value "Great" in some rows, where they should be integers.

> "Ease of Online booking" has "Unknown" where there should be integers.


Some columns ("Baggage handling", "Ease of Online booking", and "Food and drink") and entered as "object" data types, rather than float64, which appears to be how the other columns with similar information (ratings on how that metric is scored) are contained within the dataframe. 

The lack of mandatory fields leaves many rows with large amounts of nulls, with only "Class" as an available value. It becomes difficult to conduct analysis with large amounts of N/As, and having more consistency with mandatory fields would be of value in analysis. 


## Additional observations

> In the intial dataset, there were 104,910 rows and 16 columns.


> We created a different dataframe 'df_Business' which contained 50,000 rows and 16 columns.


> We also observed unique values in each column.


> There are missing values in every column.


> We also have 3,064 duplicate rows.


# Data Cleansing

## 2. Conduct Data Cleansing:

Implement the following tasks: 
 1. Drop the Class column from the created Business Class passenger satisfaction survey (“df_Business”).
 2. Rename “df_Business” columns by making all the columns’ names in lowercase, and if the column name has whitespace, replace it with an underscore. (For example: “Customer Type” will be renamed to “customer_type”).
 3. Modify improperly formatted data and handle bad data.
 4. Handle the missing values (You should have zero missing values in your cleaned dataset).
 5. Remove the fully duplicated observations.
 6. Fix the columns datatypes.

### 1. Drop the Class column from the created Business Class passenger satisfaction survey (“df_Business”).

In [18]:
# Dropping columns
df_Business.drop(['Class'], axis=1, inplace=True)
df_Business

Gender      Customer Type   Age   Type of Travel  Flight Distance  \
0         Male     Loyal Customer  69.0  Personal Travel          -5421.0   
2         Male  Disloyal Customer  41.0  Business travel              0.0   
3         Male     Loyal Customer  52.0  Business travel              0.0   
4       Female  Disloyal Customer  21.0  Business travel              0.0   
5       Female     Loyal Customer  57.0  Business travel              0.0   
...        ...                ...   ...              ...              ...   
104905    Male     Loyal Customer  45.0  Business travel           3305.0   
104906  Female     Loyal Customer  35.0  Business travel            337.0   
104907    Male     Loyal Customer  47.0  Business travel           3496.0   
104908    Male     Loyal Customer  42.0  Business travel           3605.0   
104909  Female           disloyal   0.0  Business travel            862.0   

        Inflight wifi service  Departure/Arrival time convenient  \
0                         3.0                                2.0   
2                         1.0                                1.0   
3                         2.0                                4.0   
4                         4.0                                0.0   
5                         3.0                                4.0   
...                       ...                                ...   
104905                    2.0                                2.0   
104906                    5.0                                5.0   
104907                    2.0                                2.0   
104908                    4.0                                4.0   
104909                    5.0                                5.0   

       Ease of Online booking Food and drink  Inflight entertainment  \
0                           3              5                     1.0   
2                         NaN            NaN                     NaN   
3                           4              1                     2.0   
4                         NaN            NaN                     NaN   
5                           4              5                     2.0   
...                       ...            ...                     ...   
104905                      2              3                     4.0   
104906                      5              2                     4.0   
104907                      2              2                     5.0   
104908                      4              3                     4.0   
104909                      5              5                     5.0   

       Baggage handling  Cleanliness  Departure Delay in Minutes  \
0                     1          3.0                        14.0   
2                   NaN          NaN                         NaN   
3                     2          4.0                         0.0   
4                   NaN          NaN                         NaN   
5                     3          2.0                         0.0   
...                 ...          ...                         ...   
104905                4          3.0                         8.0   
104906                4          5.0                        54.0   
104907                5          4.0                         0.0   
104908                4          3.0                         0.0   
104909                1          5.0                         9.0   

        Arrival Delay in Minutes  Satisfaction  
0                            0.0  Dissatisfied  
2                            NaN       neutral  
3                            6.0  Dissatisfied  
4                            NaN       neutral  
5                            0.0  Dissatisfied  
...                          ...           ...  
104905                       0.0     satisfied  
104906                      40.0     satisfied  
104907                       2.0     satisfied  
104908                       0.0     satisfied  
104909                       5.0       neutra

### 2. Rename “df_Business” columns by making all the columns’ names in lowercase, and if the column name has whitespace, replace it with an underscore. (For example: “Customer Type” will be renamed to “customer_type”).

In [19]:
df_Business.columns

Index(['Gender', 'Customer Type', 'Age', 'Type of Travel', 'Flight Distance',
       'Inflight wifi service', 'Departure/Arrival time convenient',
       'Ease of Online booking', 'Food and drink', 'Inflight entertainment',
       'Baggage handling', 'Cleanliness', 'Departure Delay in Minutes',
       'Arrival Delay in Minutes', 'Satisfaction'],
      dtype='object')

In [20]:
df_Business.rename(columns = lambda x: x.lower().replace(' ','_'), inplace= True)
df_Business.columns

Index(['gender', 'customer_type', 'age', 'type_of_travel', 'flight_distance',
       'inflight_wifi_service', 'departure/arrival_time_convenient',
       'ease_of_online_booking', 'food_and_drink', 'inflight_entertainment',
       'baggage_handling', 'cleanliness', 'departure_delay_in_minutes',
       'arrival_delay_in_minutes', 'satisfaction'],
      dtype='object')

In [21]:
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 104909
Data columns (total 15 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   gender                             48239 non-null  object 
 1   customer_type                      48239 non-null  object 
 2   age                                47373 non-null  float64
 3   type_of_travel                     48239 non-null  object 
 4   flight_distance                    46604 non-null  float64
 5   inflight_wifi_service              48239 non-null  float64
 6   departure/arrival_time_convenient  48239 non-null  float64
 7   ease_of_online_booking             42705 non-null  object 
 8   food_and_drink                     42705 non-null  object 
 9   inflight_entertainment             43649 non-null  float64
 10  baggage_handling                   43649 non-null  object 
 11  cleanliness                        43649 non-null  fl

### 3. Modify improperly formatted data and handle bad data.

In [22]:
# Replacing specific values 

df_Business.replace({'gender': {'F': 'Female'}}, inplace=True)
df_Business.replace({'gender': {'M': 'Male'}}, inplace=True)
df_Business.replace({'customer_type': {'disloyal': 'Disloyal Customer'}}, inplace=True)
df_Business.replace({'customer_type': {'Loyal': 'Loyal Customer'}}, inplace=True)
df_Business.replace({'type_of_travel': {'Business': 'Business travel'}}, inplace=True)
df_Business.replace({'satisfaction': {'dissatisfied': 'Dissatisfied'}}, inplace=True)
df_Business.replace({'satisfaction': {'satisfied': 'Satisfied'}}, inplace=True)
df_Business.replace({'satisfaction': {'neutral': 'Neutral'}}, inplace=True)

In [23]:
df_Business

gender      customer_type   age   type_of_travel  flight_distance  \
0         Male     Loyal Customer  69.0  Personal Travel          -5421.0   
2         Male  Disloyal Customer  41.0  Business travel              0.0   
3         Male     Loyal Customer  52.0  Business travel              0.0   
4       Female  Disloyal Customer  21.0  Business travel              0.0   
5       Female     Loyal Customer  57.0  Business travel              0.0   
...        ...                ...   ...              ...              ...   
104905    Male     Loyal Customer  45.0  Business travel           3305.0   
104906  Female     Loyal Customer  35.0  Business travel            337.0   
104907    Male     Loyal Customer  47.0  Business travel           3496.0   
104908    Male     Loyal Customer  42.0  Business travel           3605.0   
104909  Female  Disloyal Customer   0.0  Business travel            862.0   

        inflight_wifi_service  departure/arrival_time_convenient  \
0                         3.0                                2.0   
2                         1.0                                1.0   
3                         2.0                                4.0   
4                         4.0                                0.0   
5                         3.0                                4.0   
...                       ...                                ...   
104905                    2.0                                2.0   
104906                    5.0                                5.0   
104907                    2.0                                2.0   
104908                    4.0                                4.0   
104909                    5.0                                5.0   

       ease_of_online_booking food_and_drink  inflight_entertainment  \
0                           3              5                     1.0   
2                         NaN            NaN                     NaN   
3                           4              1                     2.0   
4                         NaN            NaN                     NaN   
5                           4              5                     2.0   
...                       ...            ...                     ...   
104905                      2              3                     4.0   
104906                      5              2                     4.0   
104907                      2              2                     5.0   
104908                      4              3                     4.0   
104909                      5              5                     5.0   

       baggage_handling  cleanliness  departure_delay_in_minutes  \
0                     1          3.0                        14.0   
2                   NaN          NaN                         NaN   
3                     2          4.0                         0.0   
4                   NaN          NaN                         NaN   
5                     3          2.0                         0.0   
...                 ...          ...                         ...   
104905                4          3.0                         8.0   
104906                4          5.0                        54.0   
104907                5          4.0                         0.0   
104908                4          3.0                         0.0   
104909                1          5.0                         9.0   

        arrival_delay_in_minutes  satisfaction  
0                            0.0  Dissatisfied  
2                            NaN       Neutral  
3                            6.0  Dissatisfied  
4                            NaN       Neutral  
5                            0.0  Dissatisfied  
...                          ...           ...  
104905                       0.0     Satisfied  
104906                      40.0     Satisfied  
104907                       2.0     Satisfied  
104908                       0.0     Satisfied  
104909                       5.0       Neutra

In [24]:
# Dropping rows where there is bad data in numeric columns (outliers, unknowns, etc.)

df_Business.drop(df_Business[df_Business['ease_of_online_booking'] == 'Unknown'].index, inplace = True)
df_Business.drop(df_Business[df_Business['inflight_entertainment'] == 10].index, inplace = True)
df_Business.drop(df_Business[df_Business['baggage_handling'] == 'A'].index, inplace = True)
df_Business.drop(df_Business[df_Business['satisfaction'] == 0].index, inplace = True)
df_Business.drop(df_Business[df_Business['age'] >= 180].index, inplace = True)
df_Business.drop(df_Business[df_Business['age'] <= 0].index, inplace = True)
df_Business.drop(df_Business[df_Business['flight_distance'] <= 0].index, inplace = True)

df_Business.shape

(47676, 15)

In [25]:
Column_Values = df_Business.columns

for col in Column_Values:
    print(col)
    
    unique_values = df_Business[col].unique()
    print(unique_values, '\n-------------------------------------------\n\n')

gender
['Female' 'Male' nan] 
-------------------------------------------


customer_type
['Loyal Customer' 'Disloyal Customer' nan] 
-------------------------------------------


age
[39. 33. 55. 59. 53. 56. 47. 54. 57. 25. 51. 60. 27. 43. 69. 19. 63. 50.
 nan 38. 46. 85. 45. 41. 26. 42. 30. 22. 52. 17. 48. 40. 16. 32. 35. 44.
 49. 65. 67. 66. 64.  7. 21. 37. 36. 15. 62. 34. 68. 70. 58. 11. 24. 74.
 13. 80. 75. 61. 71. 12. 73. 72. 20. 79. 10. 18. 76. 14. 77.  8.  9. 78.] 
-------------------------------------------


type_of_travel
['Business travel' 'Personal Travel' nan] 
-------------------------------------------


flight_distance
[5.600e+01 6.700e+01 7.300e+01 ... 4.983e+03 2.820e+06 9.000e+06] 
-------------------------------------------


inflight_wifi_service
[ 2.  4.  5.  3.  1.  0. nan] 
-------------------------------------------


departure/arrival_time_convenient
[ 5.  3.  4.  2.  1.  0. nan] 
-------------------------------------------


ease_of_online_booking
['5' '4' n

### 4. Handle the missing values (You should have zero missing values in your cleaned dataset).

In [26]:
# Removing missing variables from numeric data

df_Business.dropna(subset=['age', 'flight_distance', 'inflight_wifi_service','departure/arrival_time_convenient', 'ease_of_online_booking', 'food_and_drink', 'inflight_entertainment', 'baggage_handling', 'cleanliness', 'departure_delay_in_minutes', 'arrival_delay_in_minutes'], inplace=True)
df_Business.shape

(39559, 15)

In [27]:
# Replacing missing variables in the 'Satisfaction' column with 'Unknown'

df_Business['satisfaction'].fillna('Unknown', inplace=True)

In [28]:
# Number of missing values in each coulmn
df_Business.isnull().sum()

gender                               0
customer_type                        0
age                                  0
type_of_travel                       0
flight_distance                      0
inflight_wifi_service                0
departure/arrival_time_convenient    0
ease_of_online_booking               0
food_and_drink                       0
inflight_entertainment               0
baggage_handling                     0
cleanliness                          0
departure_delay_in_minutes           0
arrival_delay_in_minutes             0
satisfaction                         0
dtype: int64

### 5. Remove the fully duplicated observations.

In [29]:
# Print the number of duplicate observations
df_Business.duplicated().sum()

423

In [30]:
# Drop Duplicate Observations
df_Business.drop_duplicates(inplace=True)

In [31]:
print("Re-checking the number of duplicate observations:")
df_Business.duplicated().sum()

Re-checking the number of duplicate observations:


0

### 6. Fix the columns datatypes.

In [34]:
# Converting numeric data types to integers

df_Business = df_Business.astype({'age': 'int64','flight_distance': 'int64', 'inflight_wifi_service': 'int64', 'ease_of_online_booking': 'int64', 'food_and_drink': 'int64','inflight_entertainment': 'int64', 'baggage_handling': 'int64', 'cleanliness': 'int64', 'departure_delay_in_minutes': 'int64', 'arrival_delay_in_minutes': 'int64','departure/arrival_time_convenient': 'int64'})

df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 39136 entries, 15 to 104908
Data columns (total 15 columns):
 #   Column                             Non-Null Count  Dtype 
---  ------                             --------------  ----- 
 0   gender                             39136 non-null  object
 1   customer_type                      39136 non-null  object
 2   age                                39136 non-null  int64 
 3   type_of_travel                     39136 non-null  object
 4   flight_distance                    39136 non-null  int64 
 5   inflight_wifi_service              39136 non-null  int64 
 6   departure/arrival_time_convenient  39136 non-null  int64 
 7   ease_of_online_booking             39136 non-null  int64 
 8   food_and_drink                     39136 non-null  int64 
 9   inflight_entertainment             39136 non-null  int64 
 10  baggage_handling                   39136 non-null  int64 
 11  cleanliness                        39136 non-null  int64 
 12  de

In [35]:
# Reset the index of the DataFrame
df_Business.reset_index(drop=True, inplace = True)
df_Business

gender   customer_type  age   type_of_travel  flight_distance  \
0      Female  Loyal Customer   39  Business travel               56   
1      Female  Loyal Customer   33  Business travel               56   
2        Male  Loyal Customer   55  Business travel               67   
3      Female  Loyal Customer   56  Business travel               67   
4        Male  Loyal Customer   47  Business travel               67   
...       ...             ...  ...              ...              ...   
39131  Female  Loyal Customer   49  Business travel          9000000   
39132    Male  Loyal Customer   45  Business travel             3305   
39133  Female  Loyal Customer   35  Business travel              337   
39134    Male  Loyal Customer   47  Business travel             3496   
39135    Male  Loyal Customer   42  Business travel             3605   

       inflight_wifi_service  departure/arrival_time_convenient  \
0                          2                                  5   
1                          4                                  5   
2                          5                                  5   
3                          1                                  4   
4                          5                                  5   
...                      ...                                ...   
39131                      4                                  4   
39132                      2                                  2   
39133                      5                                  5   
39134                      2                                  2   
39135                      4                                  4   

       ease_of_online_booking  food_and_drink  inflight_entertainment  \
0                           5               5                       2   
1                           4               1                       4   
2                           5               1                       5   
3                           4               1                       3   
4                           5               3                       4   
...                       ...             ...                     ...   
39131                       2               3                       4   
39132                       2               3                       4   
39133                       5               2                       4   
39134                       2               2                       5   
39135                       4               3                       4   

       baggage_handling  cleanliness  departure_delay_in_minutes  \
0                     5            5                           0   
1                     4            2                           0   
2                     5            2                          10   
3                     1            3                           0   
4                     5            2                           0   
...                 ...          ...                         ...   
39131                 4            3                           0   
39132                 4            3                           8   
39133                 4            5                          54   
39134                 5            4                           0   
39135                 4            3                           0   

       arrival_delay_in_minutes  satisfaction  
0                             0     Satisfied  
1                             0     Satisfied  
2                            12     Satisfied  
3                             0  Dissatisfied  
4                             0     Satisfied  
...                         ...           ...  
39131                         0     Satisfied  
39132                         0     Satisfied  
39133                        40     Satisfied  
39134                         2     Satisfied  
39135                         0     Satisfied  

[39136 rows x 15 columns]

In [36]:
# Sorting column unique values
Age_uniques = df_Business['age'].unique()
sorted(Age_uniques)

[11,
 12,
 13,
 14,
 15,
 16,
 17,
 19,
 21,
 22,
 25,
 26,
 27,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 85]

In [37]:
# for each column
cols = df_Business.columns
for col in cols:
    print(col)

    # get a list of unique values
    unique = df_Business[col].unique()
    print(sorted(unique), '\n*************************************\n\n')

gender
['Female', 'Male'] 
*************************************


customer_type
['Disloyal Customer', 'Loyal Customer'] 
*************************************


age
[11, 12, 13, 14, 15, 16, 17, 19, 21, 22, 25, 26, 27, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 85] 
*************************************


type_of_travel
['Business travel', 'Personal Travel'] 
*************************************


flight_distance
[56, 67, 73, 74, 76, 77, 78, 80, 82, 83, 84, 86, 89, 90, 95, 96, 98, 100, 101, 102, 106, 107, 108, 109, 110, 112, 113, 114, 115, 116, 118, 119, 120, 121, 122, 123, 125, 126, 127, 128, 129, 130, 133, 134, 135, 137, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 150, 151, 152, 153, 154, 156, 157, 158, 159, 160, 162, 163, 164, 166, 168, 169, 170, 171, 173, 174, 175, 177, 179, 180, 181, 182, 183, 184, 185, 187, 188, 189, 190, 

# Export the cleaned dataset to CSV file.

In [38]:
# Exporting data into csv file
df_Business.to_csv("Project_2_RothEagle_Group12.csv")

# Compare the Data Characteristics Before and After Data Cleansing.

### This is the data frame (df_Business) after data cleansing. 
##### You will observe these changes:
 
> In gender: replaced “F” with “Female” and “M” with “Male”

> In “Customer Type” replaced “disloyal” with “Disloyal Customer” and “Loyal” with “Loyal Customer”

> Age: dropped nan, removed outliers '180' and 185'

> Type of travel: replaced “Business” with “Business travel”,

> Inflight wifi service: dropped nan

> Flight distance: removed outliers '-5421' and '0'

> Departure time: replace dropped nan

> Ease of online booking: removed “Unknown”, removed nan

> Food and drink: removed “Great”

> Inflight entertainment: removed “10” and removed “nan”

> Baggage handling: removed “nan” and removed “A”

> Cleanliness: removed nan

> Departure delay: removed nan

> Arrival delay: removed nan

> Satisfaction: changed “dissatisfied” to “Dissatisfied” and changed “satisfied” to “Satisfied”

# * Before Cleansing *

### Columns names and datatype

In [14]:
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 104909
Data columns (total 16 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Gender                             48239 non-null  object 
 1   Customer Type                      48239 non-null  object 
 2   Age                                47373 non-null  float64
 3   Type of Travel                     48239 non-null  object 
 4   Class                              50000 non-null  object 
 5   Flight Distance                    46604 non-null  float64
 6   Inflight wifi service              48239 non-null  float64
 7   Departure/Arrival time convenient  48239 non-null  float64
 8   Ease of Online booking             42705 non-null  object 
 9   Food and drink                     42705 non-null  object 
 10  Inflight entertainment             43649 non-null  float64
 11  Baggage handling                   43649 non-null  ob

#### Each Column name and Dataype:

|  Column                            | Dtype  
|  :------                           |   -----:
|  Gender                            |   object  
|  Customer Type                     |   object 
|  Age                               |   float64 
|  Type of Travel                    |   object 
|  Class                             |   object 
|  Flight Distance                   |   float64
|  Inflight wifi service             |   float64 
|  Departure/Arrival time convenient |   float64  
|  Ease of Online booking            |   object 
|  Food and drink                    |   object  
|  Inflight entertainment            |   float64 
|  Baggage handling                  |   object  
|  Cleanliness                       |   float64 
|  Departure Delay in Minutes        |   float64 
|  Arrival Delay in Minutes          |   float64  
|  Satisfaction                      |   object  

### Unique Values for each Column

In [15]:
df_Business.columns

Index(['Gender', 'Customer Type', 'Age', 'Type of Travel', 'Class',
       'Flight Distance', 'Inflight wifi service',
       'Departure/Arrival time convenient', 'Ease of Online booking',
       'Food and drink', 'Inflight entertainment', 'Baggage handling',
       'Cleanliness', 'Departure Delay in Minutes', 'Arrival Delay in Minutes',
       'Satisfaction'],
      dtype='object')

In [16]:
cols = df_Business.columns

# for each column
for col in cols:
    print(col)

    # get a list of unique values
    unique = df_Business[col].unique()
    print(unique, '\n---------------------------------------\n\n')

Gender
['Male' 'Female' 'F' 'M' nan] 
---------------------------------------


Customer Type
['Loyal Customer' 'Disloyal Customer' nan 'disloyal' 'Loyal'] 
---------------------------------------


Age
[ 69.  41.  52.  21.  57.  45.  39.  33.  10.  55.  59.  53.  56.  47.
  54.  25.  51.  60.  27.  43.  20.  19.  63.  50.  nan  38.  46.  85.
  26.  42.  30.  24.  22.  17.  48.  40.  16.  32.  35.  44.  49.  65.
  67.  66.  64.   7.  37.  36.  15.  62.  34.   0.  68.  70.  58.  11.
  74.  13.  80.  75.   9.  61.  71.  12.  73.  72.  18.  79.   8.  76.
  14.  77.  78. 180. 185.] 
---------------------------------------


Type of Travel
['Personal Travel' 'Business travel' 'Business' nan] 
---------------------------------------


Class
['Business'] 
---------------------------------------


Flight Distance
[-5.421e+03  0.000e+00  5.600e+01 ...  4.983e+03  2.820e+06  9.000e+06] 
---------------------------------------


Inflight wifi service
[ 3.  1.  2.  4.  0.  5. nan] 
---------------

### Number of Missing Values for each Column

In [17]:
df_Business.isnull().sum()

Gender                               1761
Customer Type                        1761
Age                                  2627
Type of Travel                       1761
Class                                   0
Flight Distance                      3396
Inflight wifi service                1761
Departure/Arrival time convenient    1761
Ease of Online booking               7295
Food and drink                       7295
Inflight entertainment               6351
Baggage handling                     6351
Cleanliness                          6351
Departure Delay in Minutes           6351
Arrival Delay in Minutes             7902
Satisfaction                         2160
dtype: int64

# *** After Cleansing ***

### Columns names and datatype

In [39]:
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39136 entries, 0 to 39135
Data columns (total 15 columns):
 #   Column                             Non-Null Count  Dtype 
---  ------                             --------------  ----- 
 0   gender                             39136 non-null  object
 1   customer_type                      39136 non-null  object
 2   age                                39136 non-null  int64 
 3   type_of_travel                     39136 non-null  object
 4   flight_distance                    39136 non-null  int64 
 5   inflight_wifi_service              39136 non-null  int64 
 6   departure/arrival_time_convenient  39136 non-null  int64 
 7   ease_of_online_booking             39136 non-null  int64 
 8   food_and_drink                     39136 non-null  int64 
 9   inflight_entertainment             39136 non-null  int64 
 10  baggage_handling                   39136 non-null  int64 
 11  cleanliness                        39136 non-null  int64 
 12  depa

### Unique Values for each Column

In [40]:
cols = df_Business.columns

# for each column
for col in cols:
    print(col)

    # get a list of unique values
    unique = df_Business[col].unique()
    print(unique, '\n---------------------------------------\n\n')

gender
['Female' 'Male'] 
---------------------------------------


customer_type
['Loyal Customer' 'Disloyal Customer'] 
---------------------------------------


age
[39 33 55 56 47 54 57 25 51 60 27 69 19 63 50 38 46 85 59 41 26 42 22 52
 17 48 16 53 32 35 45 44 49 65 67 66 64 21 37 36 15 62 40 43 68 70 58 34
 11 74 13 80 75 71 12 61 73 72 79 76 14 77 78] 
---------------------------------------


type_of_travel
['Business travel' 'Personal Travel'] 
---------------------------------------


flight_distance
[     56      67      73 ...    4983 2820000 9000000] 
---------------------------------------


inflight_wifi_service
[2 4 5 1 0 3] 
---------------------------------------


departure/arrival_time_convenient
[5 4 3 2 1 0] 
---------------------------------------


ease_of_online_booking
[5 4 0 2 3 1] 
---------------------------------------


food_and_drink
[5 1 3 4 2 0] 
---------------------------------------


inflight_entertainment
[2 4 5 3 1 0] 
---------------------------

### Number of Missing Values for each Column

In [41]:
df_Business.isnull().sum()

gender                               0
customer_type                        0
age                                  0
type_of_travel                       0
flight_distance                      0
inflight_wifi_service                0
departure/arrival_time_convenient    0
ease_of_online_booking               0
food_and_drink                       0
inflight_entertainment               0
baggage_handling                     0
cleanliness                          0
departure_delay_in_minutes           0
arrival_delay_in_minutes             0
satisfaction                         0
dtype: int64